File: wideFinder/wf.js
var fileName = process.argv[2];
var fs = require('fs');
var bufferSize = 64 * 1024 * 1024;
var size = fs.statSync(fileName).size;

function sortObject(obj) {
  var array = [], k, i;
  for (k in obj) {
    array.push({key: k, value: obj[k]});
  }
  array.sort(function(a, b) {
    return (b.value - a.value);
  });
  for (i = 0; i < 10; i += 1) {
    console.log('     ' + array[i].value + ':' + array[i].key.substring(0,59));
  }
}

function updateObject(obj, key, value) {
  if (obj[key]) {
    obj[key] += value;
  } else {
    obj[key] = value;
  }
}

function parse(lines, u_hits, u_bytes, s404s, clients, refs) {
  var i, line, f, client, u, status, bytes, ref;
  var pattern = /^\/ongoing\/When\/\d\d\dx\/\d\d\d\d\/\d\d\/\d\d\/[^\s\.]+$/;
  var refPattern = /^"http:\/\/www\.tbray\.org\/ongoing\//;
  for (i = 0; i < lines.length; i += 1) {
    line = lines[i];
    f = line.split(/\s/);
    if (f[5] === '"GET') {
      client = f[0];
      u = f[6];
      status = f[8];
      bytes = f[9];
      ref = f[10];
      if (status === '200') {
        updateObject(u_bytes, u, parseInt(bytes, 10));
        if (pattern.test(u)) {
          updateObject(u_hits, u, 1);
          updateObject(clients, client, 1);
          if (!(ref === '"-"' || refPattern.test(ref))) {
            updateObject(refs, ref.substring(1, ref.length-1), 1);
          }
        }
      } else if (status === '304') {
        if (pattern.test(u)) {
          updateObject(u_hits, u, 1);
          updateObject(clients, client, 1);
          if (!(ref === '"-"' || refPattern.test(ref))) {
            updateObject(refs, ref.substring(1, ref.length-1), 1);
          }
        }
      } else if (status === '404') {
        updateObject(s404s, u, 1);
      }
    }
  }
}

function readThrough(fd, size, bufferSize, position, prefix, process, u_hits, u_bytes, s404s, clients, refs) {
  fs.read(fd, new Buffer(bufferSize), 0, bufferSize, position, function(err, bytesRead, buffer){
    var lines = buffer.toString('utf8').split(/\n/);
    if(prefix.length > 0) {
      lines[0] = prefix + lines[0];
    }
    var last = lines.pop();
    process(lines, u_hits, u_bytes, s404s, clients, refs);
    if (position + bufferSize >= size) {
      console.log('file finished');
      console.log('URIs by hit');
      sortObject(u_hits);
      console.log('URIs by bytes');
      sortObject(u_bytes);
      console.log('404s');
      sortObject(s404s);
      console.log('client addresses');
      sortObject(clients);
      console.log('referrers');
      sortObject(refs);
      return console.log('done');
    } else {
      readThrough(fd, size, bufferSize, (position + bytesRead + 1), last, process, u_hits, u_bytes, s404s, clients, refs);
    }
  });
}

fs.open(fileName,'r', function(err, fd) {
  if (err) {
    return console.log(err.message);
  }
  readThrough(fd, size, bufferSize, 0, '', parse, {}, {}, {}, {}, {});
});

File: wideFinder/wf.rb
class Hash

  def top(howmany)
    count = 0
    keys = []
    min = 0
    self.each do |key, val|
      if val > min
        keys << key
        keys.sort! do |k1, k2|
          diff = self[k2] - self[k1]
          (diff != 0) ? diff : k1 <=> k2
        end
        keys.pop if keys.length > howmany
        min = self[keys[-1]]
      end
    end
    keys
  end
end

@u_hits = {}
@u_bytes = {}
@s404s = {}
@clients = {}
@refs = {}
@count = 0

@u_hits.default = @u_bytes.default = @s404s.default =
  @clients.default = @refs.default = 0

def record(client, u, bytes, ref)
  @u_bytes[u] += bytes
  @count += 1
  if u =~  %r{^/ongoing/When/\d\d\dx/\d\d\d\d/\d\d/\d\d/[^ .]+$}
    @u_hits[u] += 1
    @clients[client] += 1
    unless (ref == '"-"' || ref =~ %r{^\"http://www.tbray.org/ongoing/})
      @refs[ref[1 .. -2]] += 1 # lose the quotes
    end
  end
end

def report(label, hash, shrink = false)
  puts "Top #{label}:"
  keys_by_count = hash.top(10)
  fmt = (shrink) ? " %9.1fM: %s\n" : " %10d: %s\n"
  keys_by_count.each do |key|
    pkey = (key.length > 60) ? key[0 .. 59] + "..." : key
    hash[key] = hash[key] / (1024.0 * 1024.0) if shrink
    printf fmt, hash[key], pkey
  end
  puts
end

ARGF.each_line do |line|
  f = line.split(/\s+/)
  next unless f[5] == '"GET'
  client, u, status, bytes, ref = f[0], f[6], f[8], f[9], f[10]

  if status == '200'
    record(client, u, bytes.to_i, ref)
  elsif status == '304'
    record(client, u, 0, ref)
  elsif status == '404'
    @s404s[u] += 1
  end
end

print "#{@u_hits.size} resources, #{@s404s.size} 404s, #{@clients.size} clients\n\n"

report('URIs by hit', @u_hits)
report('URIs by bytes', @u_bytes, true)
report('404s', @s404s)
report('client addresses', @clients)
report('referrers', @refs)
